several vim tests fail

Vim tests fail when vim-8.1.2027 is configured as follows:
```
$ ./configure --with-features=normal --enable-gui=none --enable-terminal
$ make ; make test
...snip..
Executed:  2023 Tests
 Skipped:    47 Tests
  FAILED:     2 Tests


Failures: 
	From test_popupwin.vim:
	Found errors in Test_popup_beval():
	Run 1:
	function RunTheTest[40]..Test_popup_beval[32]..VerifyScreenDump line 55: See dump file difference: call term_dumpdiff("testdir/failed/Test_popupwin_beval_1.dump", "testdir/dumps/Test_popupwin_beval_1.dump"); difference in line 4: "|4| @73"
	function RunTheTest[40]..Test_popup_beval[35]..VerifyScreenDump line 55: See dump file difference: call term_dumpdiff("testdir/failed/Test_popupwin_beval_2.dump", "testdir/dumps/Test_popupwin_beval_2.dump"); difference in line 4: "|4| @73"
	Run 2:
	function RunTheTest[40]..Test_popup_beval[32]..VerifyScreenDump line 55: See dump file difference: call term_dumpdiff("testdir/failed/Test_popupwin_beval_1.dump", "testdir/dumps/Test_popupwin_beval_1.dump"); difference in line 4: "|4| @73"
	function RunTheTest[40]..Test_popup_beval[35]..VerifyScreenDump line 55: See dump file difference: call term_dumpdiff("testdir/failed/Test_popupwin_beval_2.dump", "testdir/dumps/Test_popupwin_beval_2.dump"); difference in line 4: "|4| @73"
	Flaky test failed too often, giving up
	From test_terminal.vim:
	Found errors in Test_terminal_termwinkey():
	Run 1:
	function RunTheTest[40]..Test_terminal_termwinkey line 22: Expected 1124 but got 1114
	function RunTheTest[40]..Test_terminal_termwinkey line 29: Expected 1124 but got 1114
	function RunTheTest[40]..Test_terminal_termwinkey line 35: Expected 1124 but got 1114
	function RunTheTest[40]..Test_terminal_termwinkey[39]..WaitForAssert[2]..<SNR>7_WaitForCommon[11]..<lambda>142 line 1: Expected 'dead' but got 'run'
	Run 2:
	function RunTheTest[40]..Test_terminal_termwinkey line 22: Expected 1127 but got 1124
	function RunTheTest[40]..Test_terminal_termwinkey line 29: Expected 1127 but got 1124
	function RunTheTest[40]..Test_terminal_termwinkey line 35: Expected 1127 but got 1124
	function RunTheTest[40]..Test_terminal_termwinkey[39]..WaitForAssert[2]..<SNR>7_WaitForCommon[11]..<lambda>143 line 1: Expected 'dead' but got 'run'
	Run 3:
	function RunTheTest[40]..Test_terminal_termwinkey line 22: Expected 1130 but got 1127
	function RunTheTest[40]..Test_terminal_termwinkey line 29: Expected 1130 but got 1127
	function RunTheTest[40]..Test_terminal_termwinkey line 35: Expected 1130 but got 1127
	function RunTheTest[40]..Test_terminal_termwinkey[39]..WaitForAssert[2]..<SNR>7_WaitForCommon[11]..<lambda>144 line 1: Expected 'dead' but got 'run'
	Run 4:
	function RunTheTest[40]..Test_terminal_termwinkey line 22: Expected 1133 but got 1130
	function RunTheTest[40]..Test_terminal_termwinkey line 29: Expected 1133 but got 1130
	function RunTheTest[40]..Test_terminal_termwinkey line 35: Expected 1133 but got 1130
	function RunTheTest[40]..Test_terminal_termwinkey[39]..WaitForAssert[2]..<SNR>7_WaitForCommon[11]..<lambda>145 line 1: Expected 'dead' but got 'run'
	Run 5:
	function RunTheTest[40]..Test_terminal_termwinkey line 22: Expected 1136 but got 1133
	function RunTheTest[40]..Test_terminal_termwinkey line 29: Expected 1136 but got 1133
	function RunTheTest[40]..Test_terminal_termwinkey line 35: Expected 1136 but got 1133
	function RunTheTest[40]..Test_terminal_termwinkey[39]..WaitForAssert[2]..<SNR>7_WaitForCommon[11]..<lambda>146 line 1: Expected 'dead' but got 'run'
	Flaky test failed too often, giving up

TEST FAILURE
Makefile:52: recipe for target 'report' failed
make: *** [report] Error 1
```

**Environment (please complete the following information):**
 - Vim-8.1.2027 (normal)
 - OS: xubuntu-18.04.3
 - Terminal: xfce4-terminal-0.8.7.4

--- a/runtime/doc/indent.txt
+++ b/runtime/doc/indent.txt
@@ -566,9 +566,15 @@ The examples below assume a 'shiftwidth' of 4.
 	      with "#" does not work.
 
 
+	PN    When N is non-zero recognize C pragmas, and indent them like any
+	      other code; does not concern other preprocessor directives.
+	      When N is zero (default): don't recognize C pragmas, treating
+	      them like every other preprocessor directive.
+
+
 The defaults, spelled out in full, are:
 	cinoptions=>s,e0,n0,f0,{0,}0,^0,L-1,:s,=s,l0,b0,gs,hs,N0,E0,ps,ts,is,+s,
-			c3,C0,/0,(2s,us,U0,w0,W0,k0,m0,j0,J0,)20,*70,#0
+			c3,C0,/0,(2s,us,U0,w0,W0,k0,m0,j0,J0,)20,*70,#0,P0
 
 Vim puts a line in column 1 if:
 - It starts with '#' (preprocessor directives), if 'cinkeys' contains '#0'.

--- a/src/nvim/buffer_defs.h
+++ b/src/nvim/buffer_defs.h
@@ -795,6 +795,7 @@ struct file_buffer {
   int b_ind_cpp_namespace;
   int b_ind_if_for_while;
   int b_ind_cpp_extern_c;
+  int b_ind_pragma;
 
   linenr_T b_no_eol_lnum;       /* non-zero lnum when last line of next binary
                                  * write should not have an end-of-line */

--- a/src/nvim/indent.c
+++ b/src/nvim/indent.c
@@ -295,13 +295,17 @@ int set_indent(int size, int flags)
 
   // Replace the line (unless undo fails).
   if (!(flags & SIN_UNDO) || (u_savesub(curwin->w_cursor.lnum) == OK)) {
+    const colnr_T old_offset = (colnr_T)(p - oldline);
+    const colnr_T new_offset = (colnr_T)(s - newline);
+
+    // this may free "newline"
     ml_replace(curwin->w_cursor.lnum, newline, false);
     if (!(flags & SIN_NOMARK)) {
       extmark_splice_cols(curbuf,
                           (int)curwin->w_cursor.lnum-1,
                           skipcols,
-                          (int)(p-oldline) - skipcols,
-                          (int)(s-newline) - skipcols,
+                          old_offset - skipcols,
+                          new_offset - skipcols,
                           kExtmarkUndo);
     }
 
@@ -311,15 +315,14 @@ int set_indent(int size, int flags)
 
     // Correct saved cursor position if it is in this line.
     if (saved_cursor.lnum == curwin->w_cursor.lnum) {
-      if (saved_cursor.col >= (colnr_T)(p - oldline)) {
+      if (saved_cursor.col >= old_offset) {
         // Cursor was after the indent, adjust for the number of
         // bytes added/removed.
-        saved_cursor.col += ind_len - (colnr_T)(p - oldline);
-
-      } else if (saved_cursor.col >= (colnr_T)(s - newline)) {
+        saved_cursor.col += ind_len - old_offset;
+      } else if (saved_cursor.col >= new_offset) {
         // Cursor was in the indent, and is now after it, put it back
         // at the start of the indent (replacing spaces with TAB).
-        saved_cursor.col = (colnr_T)(s - newline);
+        saved_cursor.col = new_offset;
       }
     }
     retval = true;

--- a/src/nvim/indent_c.c
+++ b/src/nvim/indent_c.c
@@ -1676,6 +1676,9 @@ void parse_cino(buf_T *buf)
   // Handle C++ extern "C" or "C++"
   buf->b_ind_cpp_extern_c = 0;
 
+  // Handle C #pragma directives
+  buf->b_ind_pragma = 0;
+
   for (p = buf->b_p_cino; *p; ) {
     l = p++;
     if (*p == '-') {
@@ -1747,6 +1750,7 @@ void parse_cino(buf_T *buf)
     case 'N': buf->b_ind_cpp_namespace = n; break;
     case 'k': buf->b_ind_if_for_while = n; break;
     case 'E': buf->b_ind_cpp_extern_c = n; break;
+    case 'P': buf->b_ind_pragma = n; break;
     }
     if (*p == ',')
       ++p;
@@ -1858,12 +1862,14 @@ int get_c_indent(void)
     goto laterend;
   }
 
-  /*
-   * #defines and so on always go at the left when included in 'cinkeys'.
-   */
+  // #defines and so on go at the left when included in 'cinkeys',
+  // exluding pragmas when customized in 'cinoptions'
   if (*theline == '#' && (*linecopy == '#' || in_cinkeys('#', ' ', true))) {
-    amount = curbuf->b_ind_hash_comment;
-    goto theend;
+    const char_u *const directive = skipwhite(theline + 1);
+    if (curbuf->b_ind_pragma == 0 || STRNCMP(directive, "pragma", 6) != 0) {
+      amount = curbuf->b_ind_hash_comment;
+      goto theend;
+    }
   }
 
   /*

--- a/src/nvim/testdir/test_cindent.vim
+++ b/src/nvim/testdir/test_cindent.vim
@@ -127,4 +127,40 @@ func Test_cindent_case()
   bwipe!
 endfunc
 
+func Test_cindent_pragma()
+  new
+  setl cindent ts=4 sw=4
+  setl cino=Ps
+
+  let code =<< trim [CODE]
+  {
+  #pragma omp parallel
+  {
+  #pragma omp task
+  foo();
+  # pragma omp taskwait
+  }
+  }
+  [CODE]
+
+  call append(0, code)
+  normal gg
+  normal =G
+
+  let expected =<< trim [CODE]
+  {
+	#pragma omp parallel
+	{
+		#pragma omp task
+		foo();
+		# pragma omp taskwait
+	}
+  }
+
+  [CODE]
+
+  call assert_equal(expected, getline(1, '$'))
+  enew! | close
+endfunc
+
 " vim: shiftwidth=2 sts=2 expandtab

